// lrs/output/Data.Bifunctor.Wrap/index.js
// Generated by psc version 0.9.3
"use strict";
var Control_Biapplicative = require("../Control.Biapplicative");
var Control_Biapply = require("../Control.Biapply");
var Control_Semigroupoid = require("../Control.Semigroupoid");
var Data_Bifunctor = require("../Data.Bifunctor");
var Data_Functor = require("../Data.Functor");
var Wrap = function (x) {
    return x;
};
var unwrap = function (v) {
    return v;
};
var functorWrap = function (dictBifunctor) {
    return new Data_Functor.Functor(function (f) {
        return function ($10) {
            return Wrap(Data_Bifunctor.rmap(dictBifunctor)(f)(unwrap($10)));
        };
    });
};
var bifunctorWrap = function (dictBifunctor) {
    return new Data_Bifunctor.Bifunctor(function (f) {
        return function (g) {
            return function ($11) {
                return Wrap(Data_Bifunctor.bimap(dictBifunctor)(f)(g)(unwrap($11)));
            };
        };
    });
};
var biapplyWrap = function (dictBiapply) {
    return new Control_Biapply.Biapply(function () {
        return bifunctorWrap(dictBiapply["__superclass_Data.Bifunctor.Bifunctor_0"]());
    }, function (v) {
        return function (v1) {
            return Control_Biapply.biapply(dictBiapply)(v)(v1);
        };
    });
};
var biapplicativeWrap = function (dictBiapplicative) {
    return new Control_Biapplicative.Biapplicative(function () {
        return biapplyWrap(dictBiapplicative["__superclass_Control.Biapply.Biapply_0"]());
    }, function (a) {
        return function (b) {
            return Control_Biapplicative.bipure(dictBiapplicative)(a)(b);
        };
    });
};
module.exports = {
    Wrap: Wrap, 
    unwrap: unwrap, 
    bifunctorWrap: bifunctorWrap, 
    functorWrap: functorWrap, 
    biapplyWrap: biapplyWrap, 
    biapplicativeWrap: biapplicativeWrap
};
